### PR TITLE
#202: Add MatchSimilarity entity and save fields similarity percentages

### PR DESCRIPTION
**Pull request by awalkowiak**

### CODE REVIEW

File: src/main/java/org/benetech/servicenet/config/CacheConfiguration.java [hunk 1]
```diff
@@ -71,6 +71,8 @@ public JCacheManagerCustomizer cacheManagerCustomizer() {
             cm.createCache(org.benetech.servicenet.domain.Program.class.getName() + ".services",
                 jcacheConfiguration);
             cm.createCache(org.benetech.servicenet.domain.ServiceAtLocation.class.getName(), jcacheConfiguration);
+            cm.createCache(org.benetech.servicenet.domain.ServiceAtLocation.class.getName() + ".phones",
```
Comment (pmuchowski) [inline src/main/java/org/benetech/servicenet/config/CacheConfiguration.java]: this was remove in one of the previous PRs

File: src/main/java/org/benetech/servicenet/matching/counter/AbstractSimilarityCounter.java [hunk 2]
```diff
@@ -15,6 +17,10 @@
      */
     public abstract float countSimilarityRatio(V obj1, V obj2, MatchingContext context);
 
+    public List getMatchSimilarityDTOs(V obj1, V obj2, MatchingContext context) {
```
Comment (pmuchowski) [inline src/main/java/org/benetech/servicenet/matching/counter/AbstractSimilarityCounter.java]: use List<MatchSimilarityDTO>

File: src/main/java/org/benetech/servicenet/matching/counter/OrganizationSimilarityCounter.java [hunk 3]
```diff
@@ -36,30 +40,98 @@
 
     @Override
     public float countSimilarityRatio(Organization org1, Organization org2, MatchingContext context) {
-        float result = 0;
-        result += nameSimilarityCounter.countSimilarityRatio(org1.getName(), org2.getName(), context)
+        float start = 0;
```
Comment (pmuchowski) [inline src/main/java/org/benetech/servicenet/matching/counter/OrganizationSimilarityCounter.java]: I don't think you need a variable for this, you can use the value directly

File: src/main/java/org/benetech/servicenet/matching/counter/OrganizationSimilarityCounter.java [hunk 4]
```diff
@@ -36,30 +40,98 @@
...
+        ));
+        similarityDtos.add(getFieldMatchSimilarityDTO(
+            getUrlSimilarity(org1, org2, context),
+            "Url"
+        ));
+        similarityDtos.add(getFieldMatchSimilarityDTO(
+            getYearIncorporatedSimilarity(org1, org2, context),
+            "YearIncorporated"
+        ));
+        float start = 0;
```
Comment (pmuchowski) [inline src/main/java/org/benetech/servicenet/matching/counter/OrganizationSimilarityCounter.java]: as above

File: src/main/java/org/benetech/servicenet/service/impl/OrganizationMatchServiceImpl.java [hunk 5]
```diff
@@ -269,10 +278,19 @@ private void removeMatches(List<OrganizationMatch>  matches) {
...
+            similarityDTO.setOrganizationMatchId(mirrorMatch.getId());
+            matchSimilarityService.save(similarityDTO);
+        }
         return matches;
     }
 
-    private boolean isSimilar(Organization organization, Organization partner, MatchingContext context) {
-        return organizationSimilarityCounter.countSimilarityRatio(organization, partner, context) >= orgMatchThreshold;
+    private boolean isSimilar(List<MatchSimilarityDTO> similarityDTOS) {
+        float start = 0;
```
Comment (pmuchowski) [inline src/main/java/org/benetech/servicenet/service/impl/OrganizationMatchServiceImpl.java]: you can use the value directly

File: src/main/java/org/benetech/servicenet/service/impl/OrganizationMatchServiceImpl.java [hunk 6]
```diff
@@ -269,10 +278,19 @@ private void removeMatches(List<OrganizationMatch>  matches) {
         matches.add(save(match));
         matches.add(save(mirrorMatch));
 
+        for (MatchSimilarityDTO similarityDTO : similarityDTOS) {
+            similarityDTO.setOrganizationMatchId(match.getId());
+            matchSimilarityService.save(similarityDTO);
+            similarityDTO.setOrganizationMatchId(mirrorMatch.getId());
+            matchSimilarityService.save(similarityDTO);
```
Comment (pmuchowski) [inline src/main/java/org/benetech/servicenet/service/impl/OrganizationMatchServiceImpl.java]: why it's saved twice?

File: src/main/resources/config/liquibase/changelog/20190709132431_added_entity_MatchSimilarity.xml [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+    <!--
+        Added the entity MatchSimilarity.
+    -->
+    <changeSet id="20190709132431-1" author="jhipster">
+        <createTable tableName="match_similarity">
+            <column name="id" type="${uuidType}" defaultValueComputed="${uuidFunction}">
+                <constraints primaryKey="true" nullable="false"/>
+            </column>
+            <column name="similarity" type="${floatType}">
```
Comment (pmuchowski) [inline src/main/resources/config/liquibase/changelog/20190709132431_added_entity_MatchSimilarity.xml]: I don't see floatType property in the migration

File: src/main/resources/config/liquibase/changelog/20190709132431_added_entity_MatchSimilarity.xml [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+            <column name="organization_match_id" type="${uuidType}">
+                <constraints nullable="true" />
+            </column>
+
+            <!-- jhipster-needle-liquibase-add-column - JHipster will add columns here, do not remove-->
+        </createTable>
+
+    </changeSet>
+
+    <changeSet id="20190709132431-1-relations" author="jhipster">
```
Comment (pmuchowski) [inline src/main/resources/config/liquibase/changelog/20190709132431_added_entity_MatchSimilarity.xml]: why there is an empty chnageset?

File: src/main/java/org/benetech/servicenet/service/impl/OrganizationMatchServiceImpl.java (hunk 6 above)
Comment (awalkowiak) [reply src/main/java/org/benetech/servicenet/service/impl/OrganizationMatchServiceImpl.java]: For match and mirrorMatch (as we discussed, it is better to keep both versions - same as OrganizationMatch)

Comment (pmuchowski) [reply src/main/java/org/benetech/servicenet/service/impl/OrganizationMatchServiceImpl.java]: ok

**Comment by awalkowiak** [comment]
@pmuchowski done

File: src/main/java/org/benetech/servicenet/matching/counter/AbstractSimilarityCounter.java [hunk 9]
```diff
@@ -2,7 +2,9 @@
 
 import java.util.Collections;
 import java.util.List;
+import org.benetech.servicenet.domain.MatchSimilarity;
```
Comment (pmuchowski) [inline src/main/java/org/benetech/servicenet/matching/counter/AbstractSimilarityCounter.java]: is it used anywhere?

**Pull request merged**